File: 0037-sudoku-solver/0037-sudoku-solver.java
class Solution {
  public boolean solved = false;
  public HashSet<?>[] rows = new HashSet<?>[9];
  public HashSet<?>[] cols = new HashSet<?>[9];
  public HashSet<?>[] subs = new HashSet<?>[9];
  
  public void backtrack(int r, int c, char[][] board) {
    if (r == 9) {
      solved = true;
      return;
    }
    
    int nextR = r + (int) Math.floor((c + 1) / (double) 9);
    int nextC = (c + 1) % 9;
    
    char cell = board[r][c];
    if (cell != '.') {
      backtrack(nextR, nextC, board);
    } else {
      for (int d = 1; d <= 9; d++) {
        int subIdx = 3 * (int) Math.floor(r / (double) 3) + (int) Math.floor(c / (double) 3);
        Set rowSet = (HashSet<Integer>) rows[r];
        Set colSet = (HashSet<Integer>) cols[c];
        Set subSet = (HashSet<Integer>) subs[subIdx];
        if (!(rowSet.contains(d) || colSet.contains(d) || subSet.contains(d))) {
          rowSet.add(d);
          colSet.add(d);
          subSet.add(d);
          board[r][c] = (char) (d + '0');
          
          backtrack(nextR, nextC, board);
          
          if (!solved) {
            rowSet.remove(d);
            colSet.remove(d);
            subSet.remove(d);
            board[r][c] = '.';
          }
        }
      }
    }
  }
  
  public void solveSudoku(char[][] board) {
    for (int i = 0; i < 9; i++) {
      rows[i] = new HashSet<Integer>();
      cols[i] = new HashSet<Integer>();
      subs[i] = new HashSet<Integer>();
    }
    for (int r = 0; r < 9; r++) {
      for (int c = 0; c < 9; c++) {
        char cell = board[r][c];
        if (cell != '.') {
          int d = cell - '0';
          int subIdx = 3 * (int) Math.floor(r / (double) 3) + (int) Math.floor(c / (double) 3);
          ((HashSet<Integer>) rows[r]).add(d);
          ((HashSet<Integer>) cols[c]).add(d);
          ((HashSet<Integer>) subs[subIdx]).add(d);
        }
      }
    }

    backtrack(0, 0, board);
  }
}

File: 0037-sudoku-solver/0037-sudoku-solver.go
var (
  solved bool
  rows []map[byte]bool
  cols []map[byte]bool
  subs []map[byte]bool
)

func solveSudoku(board [][]byte)  {
  solved = false
  rows = make([]map[byte]bool, 9)
  cols = make([]map[byte]bool, 9)
  subs = make([]map[byte]bool, 9)
  
  for i:=0;i<9;i++ {
    rows[i] = map[byte]bool{}
    cols[i] = map[byte]bool{}
    subs[i] = map[byte]bool{}
  }
  for r:=0;r<9;r++ {
    for c:=0;c<9;c++ {
      cell := board[r][c]
      if cell != []byte(".")[0] {
        subIdx := 3*(r/3)+c/3
        rows[r][cell], cols[c][cell], subs[subIdx][cell] = true, true, true
      }
    }
  }
  
  backtrack(0, 0, board)
}

func backtrack(r int, c int, board [][]byte)  {
  if r == 9 {
    solved = true
    return
  }
  
  nextR, nextC, cell := r+(c+1)/9, (c+1)%9, board[r][c]
  if cell != []byte(".")[0] {
    backtrack(nextR, nextC, board)
  } else {
    for b:=[]byte("1")[0];b<=[]byte("9")[0];b++ {
      subIdx := 3*(r/3)+c/3
      _, rowsOK := rows[r][b]
      _, colsOK := cols[c][b]
      _, subsOK := subs[subIdx][b]
      if !(rowsOK || colsOK || subsOK) {
        rows[r][b] = true
        cols[c][b] = true
        subs[subIdx][b] = true
        board[r][c] = b
        
        backtrack(nextR, nextC, board)
        
        if !solved {
          delete(rows[r], b)
          delete(cols[c], b)
          delete(subs[subIdx], b)
          board[r][c] = []byte(".")[0]
        }
      }
    }
  }
}